numba/cuda/cudaimpl.py

```diff
@@ -492,6 +492,141 @@ def ptx_fp16_binary(context, builder, sig, args):
 lower_fp16_binary(stubs.fp16.hmul, 'mul')
 
 
+@lower(stubs.fp16.hdiv, types.float16, types.float16)
+def lower_fp16_divide(context, builder, sig, args):
+    arg1 = args[0]
+    arg2 = args[1]
+    # Floating point 16 division algorithm implemented
+    # using Newton Rapson algorithm.
+
+    # x = arg1, y = arg2
+    # Implement x/y
+
+    # Convert fp16 args to fp32 so we can implement
+    # division by reciprocal multiplication on fp32
+    # values
+    cvt2float_fnty = ir.FunctionType(ir.FloatType(), [ir.IntType(16)])
+    cvt2float_asm = ir.InlineAsm(cvt2float_fnty,
+                                 "cvt.f32.f16 $0, $1;",
+                                 "=f,h")
+    arg1_fp32 = builder.call(cvt2float_asm, [arg1])
+    arg2_fp32 = builder.call(cvt2float_asm, [arg2])
+
+    rcp_fnty = ir.FunctionType(ir.FloatType(), [ir.FloatType()])
+
+    # Obtain a gross estimate of the reciprocal of argument 2 (y)
+    # which will be refined. This is the first step in the Newton
+    # Rapson method.
+
+    # y0 = 1/y * (1-err), here y0 is not a precise 1/y
+    # it is an approximation and the term 'err' stands for the
+    # relative error of the approximation.
+
+    rcp_asm = ir.InlineAsm(rcp_fnty,
+                           "rcp.approx.ftz.f32 $0, $1;",
+                           "=f,f")
+    arg2_rcp = builder.call(rcp_asm, [arg2_fp32])
+
+    # Implement fp32 division via multiplication by
+    # reciprocal of arg2
+
+    # Our objective is to get a quotient x/y, correctly rounded to nearest
+    # even in fp16 format, so we can do q0 = x * yn.
+
+    # We need to perform one more Newton iteration to eliminate the error
+    # from the multiplication.
+
+    # Mathematically we have: q0 = x * 1/y * (1 - err)
+    # So err (correction_term) = x - y*q0 = fma(-y, q0, x)
+    # And q1 = err*yn + q0 = fma(err, yn, q0)
```

Comment:
These comments seems out of place. They describe steps later in the process. The first part is conditional on if the result from the first iteration has denormals. And the second part is duplicate of what's mentioned below. I suggest we can remove this part.